src/ray/gcs/redis_module/ray_redis_module.cc

```diff
@@ -171,26 +172,24 @@ int PublishTableAdd(RedisModuleCtx *ctx, RedisModuleString *pubsub_channel_str,
     return RedisModule_ReplyWithError(ctx, "error during PUBLISH");
   }
 
+  std::string notification_key;
+  REPLY_AND_RETURN_IF_NOT_OK(GetBroadcastKey(ctx, pubsub_channel_str, id, &notification_key));
   // Publish the data to any clients who requested notifications on this key.
-  RedisModuleKey *notification_key;
-  REPLY_AND_RETURN_IF_NOT_OK(OpenBroadcastKey(&notification_key, ctx, pubsub_channel_str,
-                                              id, REDISMODULE_READ | REDISMODULE_WRITE));
-  if (RedisModule_KeyType(notification_key) != REDISMODULE_KEYTYPE_EMPTY) {
-    // NOTE(swang): Sets are not implemented yet, so we use ZSETs instead.
-    REPLY_AND_RETURN_IF_FALSE(RedisModule_ZsetFirstInScoreRange(
-                                  notification_key, REDISMODULE_NEGATIVE_INFINITE,
-                                  REDISMODULE_POSITIVE_INFINITE, 1, 1) == REDISMODULE_OK,
-                              "Unable to initialize zset iterator");
-    for (; !RedisModule_ZsetRangeEndReached(notification_key);
-         RedisModule_ZsetRangeNext(notification_key)) {
-      RedisModuleString *client_channel =
-          RedisModule_ZsetRangeCurrentElement(notification_key, NULL);
+  auto it = notification_map.find(notification_key);
+  if (it != notification_map.end()) {
+    for (const std::string& client_channel : it->second) {
+      // RedisModule_Call seems to be broken and cannot accept "bb",
+      // therefore we construct a temporary redis string here, which
+      // will be garbage collected by redis.
+      auto channel = RedisModule_CreateString(
+          ctx, client_channel.data(), client_channel.size());
```

Comment:
@pcmoritz Can we delete this temporary redis string immediately after PUBLISH command finishes? In our system we do not enable redis LRU by default. That may cause redis memory leak.

Reply:
Are you talking about LRU eviction or about automatic memory management? We shouldn’t have to free strings or close keys because redis automatic memory management takes care of that. If you do need to free strings then we you would need to do it in many places.

Reply:
Sorry, I thought it was a key... It is just a Redis String, which is fine.